Document formal script syntax (#5336)
nextflow-io/nextflow: docs/config.md

## Syntax

The Nextflow configuration syntax is based on the Nextflow script syntax. It is designed for setting configuration options in a declarative manner while also allowing for dynamic expressions where appropriate.

A Nextflow config file may consist of any number of *assignments*, *blocks*, and *includes*. Config files may also contain comments in the same manner as scripts.

See {ref}`syntax-page` for more information about the Nextflow script syntax.

### Assignments

A config assignment consists of a config option and an expression separated by an equals sign:

```groovy
workDir = 'work'
docker.enabled = true
process.maxErrors = 10
```

A config option consists of an *option name* prefixed by any number of *scopes* separated by dots. Config scopes are used to group related config options. See {ref}`config-options` for the full set of config options.

The expression is typically a literal value such as a number, boolean, or string. However, any expression can be used:

```groovy
params.helper_file = "${projectDir}/assets/helper.txt"
```

### Blocks

> A config scope can also be specified as a block, allowing multiple configuration options to be set within that block. For example:

Reviewer: ```suggestion
A config scope can also be specified as a block and allows multiple configuration options to be set within that block. For example:
```

```groovy
// dot syntax
docker.enabled = true
docker.runOptions = '-u $(id -u):$(id -g)'

// block syntax
docker {
    enabled = true
    runOptions = '-u $(id -u):$(id -g)'
}
```

As a result, deeply nested config options can be assigned in various ways. For example, the following three assignments are equivalent:

```groovy
executor.retry.maxAttempt = 5

executor {
    retry.maxAttempt = 5
}

executor {
    retry {
        maxAttempt = 5
    }
}
```

### Includes

> A config file can include any number of other config files using the `includeConfig` keyword:

Reviewer 2: Is not `configuration` better here?

Author: I guess either one is fine, but it's verbose to write configuration all the time, so we tend to write configuration the first time and config there after

Reviewer 2: True that verbose, but this is a formal docs, "configuration" sounds better imo

Author: okay

Author: ```suggestion
A configuration file can include any number of other configuration files using the `includeConfig` keyword:
```

includeConfig 'path/foo.config'
```

> When a relative path is used, it is resolved against the location of the including file.

Reviewer: ```suggestion
A relative path is resolved against the location of the including file.
```
